src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -1055,23 +1082,40 @@ private CachedCodeEntry CacheCode(CachedCodeEntryKey key)
             lock (s_livecode)
             {
                 // first look for it in the cache and move it to the head
-                for (LinkedListNode<CachedCodeEntry> current = s_livecode.First; current != null; current = current.Next)
+                s_livecode.TryGetValue(key, out var entry);
+                if (entry != null)
                 {
-                    if (current.Value._key == key)
-                    {
-                        s_livecode.Remove(current);
-                        s_livecode.AddFirst(current);
-                        return current.Value;
-                    }
+                    CachePromoteLinkedListEntry(entry);
+                    return entry;
                 }
-
                 // it wasn't in the cache, so we'll add a new one.  Shortcut out for the case where cacheSize is zero.
-                if (s_cacheSize != 0)
+                else if (s_cacheSize != 0)
                 {
                     newcached = new CachedCodeEntry(key, capnames, capslist, _code, caps, capsize, _runnerref, _replref);
-                    s_livecode.AddFirst(newcached);
-                    if (s_livecode.Count > s_cacheSize)
-                        s_livecode.RemoveLast();
+                    s_livecode.Add(key, newcached);
+                    // put in linked list:
+                    if (s_livecode_first != null)
+                    {
+                        s_livecode_first._next = newcached;
+                        newcached._previous = s_livecode_first;
+                    }
+                    s_livecode_first = newcached;
+                    if (s_livecode_last == null)
+                    {
+                        s_livecode_last = newcached;
+                    }
+                    else
+                    {
+                        if (s_livecode.Count > s_cacheSize)
+                        {
+                            var last = s_livecode_last;
+                            s_livecode.Remove(last._key);
+
+                            last._next._previous = null;
+                            s_livecode_last = last._next;
+                            last._next = null;
+                        }
+                    }
```

Comment:
Nit, on CachedCodeEntryKey, you could reorder` operator ==` to test pattern first, then options, then culture, guessing that pattern is most likely to vary and culture the least.

It's probably lost in the noise, but if you assume it's very unlikely that folks will have the same pattern with different cultures, you could remove culture from the GetHashCode() implementation. I assume it doesn't show up in the profile so it's not worth it.

Reply:
Another micro optimization: since the Dictionary will use IEquatable, it will be calling Equals (rather than operator==). You could therefore put the implementation into Equals, and have operator== call it, saving a function call. It might be inlining it already of cousre.

Reply:
> if you assume it's very unlikely that folks will have the same pattern with different culture

Please don't assume that! One of our partner teams had a huge incident because of that. In Web scenarios it's likely to store the same pattern with a different culture.

Reply:
Ack - you're right. That is true.

Reply:
You can still do the other 2 micro optimizations above: reorder equals checks and move implemenation out of ==